src/shogun/evaluation/Calibration.cpp

```diff
@@ -0,0 +1,244 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Written (W) 2011-2012 Heiko Strathmann
+ * Copyright (C) 2011 Berlin Institute of Technology and Max-Planck-Society
+ */
+
+#include <shogun/evaluation/Calibration.h>
+#include <shogun/evaluation/CalibrationMethod.h>
+#include <shogun/lib/config.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/mathematics/Math.h>
+#include <shogun/mathematics/Statistics.h>
+
+using namespace shogun;
+
+CBinaryLabels* CCalibration::apply_binary(CFeatures* features)
+{
+
+	CBinaryLabels* result_labels = (CBinaryLabels*)m_machine->apply(features);
+	CCalibrationMethod* method =
+	    (CCalibrationMethod*)m_calibration_machines->get_element(0);
+	SGVector<float64_t> confidence_values =
+	    method->apply_binary(result_labels->get_values());
+	result_labels->set_values(confidence_values);
+
+	return result_labels;
+}
+
+CMulticlassLabels* CCalibration::apply_multiclass(CFeatures* features)
+{
+	index_t num_calibration_machines =
+	    m_calibration_machines->get_num_elements();
+	CMulticlassLabels* result_labels =
+	    (CMulticlassLabels*)m_machine->apply(features);
+	for (index_t i = 0; i < num_calibration_machines; ++i)
+	{
+		CCalibrationMethod* method =
+		    (CCalibrationMethod*)m_calibration_machines->get_element(i);
+		SGVector<float64_t> confidence_values =
+		    method->apply_binary(result_labels->get_multiclass_confidences(i));
+		result_labels->set_multiclass_confidences(i, confidence_values);
+	}
+
+	SGVector<float64_t> temp_confidences;
+	index_t num_classes = result_labels->get_num_classes();
+	index_t num_samples;
+
+	// normalize the probabilities
+	for (index_t i = 0; i < num_classes; ++i)
```

Comment:
all of this can be done in a vectorised fashion using linalg, see e.g. https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/preprocessor/KernelPCA.cpp#L92

Reply:
Yes, I'm currently trying to vectorize wherever possible.